### PR TITLE
fix UnixChannelUtil#isBufferCopyNeededForWrite

### PR DESCRIPTION
**Pull request by tonyshenkk**
fix not execute unit test in transport-native-unix-common-tests module

Motivation:
- Commit 047da11 introduced an bug for still copy byteBuf for composed of n(n <= IOV_MAX) NIO direct buffers
- Commit 3c4dfed add UnixChannelUtilTest in transport-native-unix-common-tests module, but not execute in maven compile

as issue #6825, #6870

Modifications:
- modified UnixChannelUtil#isBufferCopyNeededForWrite(ByteBuf), and UnixChannelUtilTest
- move UnixChannelUtilTest into transport-native-unix-common module, and add packet scope method UnixChannelUtil#isBufferCopyNeededForWrite(ByteBuf, int)

Result:
- no copy byteBuf for composed of n(n <= IOV_MAX) NIO direct buffers
- auto execute unit tests in UnixChannelUtilTest and it is easier to mock IOV_MAX

### CODE REVIEW

**Comment by normanmaurer** [comment]
@tonyshenkk please check the test-failures https://garage.netty.io/teamcity/viewLog.html?buildId=21290&buildTypeId=pulls_netty&guest=1

**Comment by tonyshenkk** [comment]
I have fix this unit test failures and pull a patch.

File: transport-native-unix-common/src/main/java/io/netty/channel/unix/Limits.java [hunk 1]
```diff
@@ -21,11 +21,22 @@
 import static io.netty.channel.unix.LimitsStaticallyReferencedJniMethods.uioMaxIov;
 
 public final class Limits {
-    public static final int IOV_MAX = iovMax();
```
Comment (Scottmitch) [inline transport-native-unix-common/src/main/java/io/netty/channel/unix/Limits.java]: can you explain the rational behind changes in this file?

**Comment by Scottmitch** [comment]
> Commit 7460d90 introduced an bug for still copy byteBuf for composed of n(n <= IOV_MAX) NIO direct buffers

Are you sure you are referencing the correct commit, and if so can you clarify why this change had any impacts on copying of ByteBufs?

File: transport-native-unix-common/src/main/java/io/netty/channel/unix/Limits.java (hunk 1 above)
Comment (tonyshenkk) [inline transport-native-unix-common/src/main/java/io/netty/channel/unix/Limits.java]: motivation:
It is very difficult to static final native method for unit test. so, I modified the static final fields in Limits to static method.

there is an unit test in io.netty.channel.unix.tests.UnixChannelUtilTest.  After modiified, it becomes very easy to do unit test, and it looks like  same efficiency as before.

I mentioned it in commit log. I confused this log doesn't show on this pull request. Maybe it is  I am not familiar with the use of github.

**Comment by tonyshenkk** [comment]
@Scottmitch  Sorry for this error commit reference.

The correct commit reference is 047da1108674f0c425f0f00d159cdec41d697268, and discuss history post at https://github.com/netty/netty/pull/6825#issuecomment-309644968

File: transport-native-unix-common-tests/pom.xml [hunk 2]
```diff
@@ -45,5 +46,40 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.powermock</groupId>
+      <artifactId>powermock-api-mockito</artifactId>
+      <version>${powermock.version}</version>
+    </dependency>
   </dependencies>
+
+  <build>
+    <testSourceDirectory>src/main/java/io/netty/channel/unix/tests</testSourceDirectory>
```
Comment (johnou) [inline transport-native-unix-common-tests/pom.xml]: Why not repackage the test class to src/test/java..?

File: transport-native-unix-common/src/main/java/io/netty/channel/unix/UnixChannelUtil.java [hunk 3]
```diff
@@ -29,6 +29,6 @@ private UnixChannelUtil() {
      * (We check this because otherwise we need to make it a new direct buffer.)
      */
     public static boolean isBufferCopyNeededForWrite(ByteBuf byteBuf) {
-        return !byteBuf.hasMemoryAddress() || !byteBuf.isDirect() || byteBuf.nioBufferCount() > IOV_MAX;
+        return !(byteBuf.hasMemoryAddress() || (byteBuf.isDirect() && byteBuf.nioBufferCount() <= getIovMax()));
```
Comment (Scottmitch) [inline transport-native-unix-common/src/main/java/io/netty/channel/unix/UnixChannelUtil.java]: consider minor refactoring to remove the outer negation:

```java
return !byteBuf.hasMemoryAddress() && (!byteBuf.isDirect() || byteBuf.nioBufferCount() > getIovMax());
```

File: pom.xml [hunk 4]
```diff
@@ -202,6 +202,8 @@
     <testJavaHome>${env.JAVA_HOME}</testJavaHome>
     <skipOsgiTestsuite>false</skipOsgiTestsuite>
     <skipAutobahnTestsuite>false</skipAutobahnTestsuite>
+    <!--avoid below packet indexed twice-->
```
Comment (Scottmitch) [inline pom.xml]: `avoid below packet indexed twice` -> should this say `package`? Also if it is excluded, how can we be sure it runs at least once?

File: transport-native-unix-common-tests/pom.xml [hunk 5]
```diff
@@ -45,5 +46,40 @@
...
+      <version>1.10.19</version>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.javassist</groupId>
+      <artifactId>javassist</artifactId>
+      <version>3.19.0-GA</version>
+    </dependency>
+    <dependency>
+      <groupId>org.powermock</groupId>
```
Comment (Scottmitch) [inline transport-native-unix-common-tests/pom.xml]: please use `mockito` instead of introducing yet another mocking framework.

File: transport-native-unix-common/src/main/java/io/netty/channel/unix/Limits.java (hunk 1 above)
Comment (Scottmitch) [inline transport-native-unix-common/src/main/java/io/netty/channel/unix/Limits.java]: > and it looks like same efficiency as before.

Unless you can demonstrate performance isn't impacted please revert this portion of the change. A potential alternative maybe to introduce an additional method we can use in tests:

```java
public static boolean isBufferCopyNeededForWrite(ByteBuf byteBuf) {
  return isBufferCopyNeededForWrite(byteBuf, IOV_MAX);
}

static boolean isBufferCopyNeededForWrite(ByteBuf byteBuf, int iovMax) {
 return ...;
}
```

File: transport-native-unix-common-tests/pom.xml [hunk 6]
```diff
@@ -46,4 +46,16 @@
       <scope>compile</scope>
     </dependency>
   </dependencies>
+
+  <build>
```
Comment (normanmaurer) [inline transport-native-unix-common-tests/pom.xml]: @tonyshenkk why is this needed ? I think we should just remove this one.

File: transport-native-unix-common/src/test/java/io/netty/channel/unix/UnixChannelUtilTest.java [hunk 7]
```diff
@@ -44,25 +45,25 @@ public void testUnPooledAllocatorIsBufferCopyNeededForWrite() {
...
-        assertCompositeByteBufIsBufferCopyNeededForWrite(alloc, 2, 0, true);
+        assertCompositeByteBufIsBufferCopyNeededForWrite(alloc, 2, 0, false);
         assertCompositeByteBufIsBufferCopyNeededForWrite(alloc, IOV_MAX + 1, 0, true);
         assertCompositeByteBufIsBufferCopyNeededForWrite(alloc, 0, 2, true);
         assertCompositeByteBufIsBufferCopyNeededForWrite(alloc, 1, 1, true);
     }
 
-    private static void assertCompositeByteBufIsBufferCopyNeededForWrite(ByteBufAllocator alloc, int numDirect,
-                                                                         int numHeap, boolean expected) {
+    private void assertCompositeByteBufIsBufferCopyNeededForWrite(ByteBufAllocator alloc, int numDirect,
```
Comment (normanmaurer) [inline transport-native-unix-common/src/test/java/io/netty/channel/unix/UnixChannelUtilTest.java]: make static again ?

Comment (tonyshenkk) [inline transport-native-unix-common/src/test/java/io/netty/channel/unix/UnixChannelUtilTest.java]: done

File: transport-native-unix-common-tests/pom.xml (hunk 6 above)
Comment (tonyshenkk) [inline transport-native-unix-common-tests/pom.xml]: yeah, you are right.  There is only need delete the property 'maven.deploy.skip'. and this was corrected.

File: transport-native-unix-common/src/test/java/io/netty/channel/unix/UnixChannelUtilTest.java [hunk 8]
```diff
@@ -46,23 +47,23 @@ public void testUnPooledAllocatorIsBufferCopyNeededForWrite() {
...
-        assertCompositeByteBufIsBufferCopyNeededForWrite(alloc, 2, 0, true);
+        assertCompositeByteBufIsBufferCopyNeededForWrite(alloc, 2, 0, false);
         assertCompositeByteBufIsBufferCopyNeededForWrite(alloc, IOV_MAX + 1, 0, true);
         assertCompositeByteBufIsBufferCopyNeededForWrite(alloc, 0, 2, true);
         assertCompositeByteBufIsBufferCopyNeededForWrite(alloc, 1, 1, true);
     }
 
     private static void assertCompositeByteBufIsBufferCopyNeededForWrite(ByteBufAllocator alloc, int numDirect,
-                                                                         int numHeap, boolean expected) {
+                                                                  int numHeap, boolean expected) {
```
Comment (Scottmitch) [inline transport-native-unix-common/src/test/java/io/netty/channel/unix/UnixChannelUtilTest.java]: revert this change?

File: transport-native-unix-common/src/test/java/io/netty/channel/unix/UnixChannelUtilTest.java [hunk 9]
```diff
@@ -46,23 +47,23 @@ public void testUnPooledAllocatorIsBufferCopyNeededForWrite() {
...
-        assertFalse(isBufferCopyNeededForWrite(byteBuf));
-        assertFalse(isBufferCopyNeededForWrite(byteBuf.asReadOnly()));
+        assertFalse(UnixChannelUtil.isBufferCopyNeededForWrite(byteBuf, IOV_MAX));
+        assertFalse(UnixChannelUtil.isBufferCopyNeededForWrite(byteBuf.asReadOnly(), IOV_MAX));
         assertTrue(byteBuf.release());
 
         byteBuf = alloc.heapBuffer();
-        assertTrue(isBufferCopyNeededForWrite(byteBuf));
-        assertTrue(isBufferCopyNeededForWrite(byteBuf.asReadOnly()));
+        assertTrue(UnixChannelUtil.isBufferCopyNeededForWrite(byteBuf, IOV_MAX));
```
Comment (Scottmitch) [inline transport-native-unix-common/src/test/java/io/netty/channel/unix/UnixChannelUtilTest.java]: consider just using the static import which was here before this PR.

**Comment by normanmaurer** [comment]
@tonyshenkk will pull in once you addressed @Scottmitch s comment.

File: transport-native-unix-common/src/test/java/io/netty/channel/unix/UnixChannelUtilTest.java (hunk 9 above)
Comment (tonyshenkk) [inline transport-native-unix-common/src/test/java/io/netty/channel/unix/UnixChannelUtilTest.java]: done

File: transport-native-unix-common/src/test/java/io/netty/channel/unix/UnixChannelUtilTest.java (hunk 8 above)
Comment (tonyshenkk) [inline transport-native-unix-common/src/test/java/io/netty/channel/unix/UnixChannelUtilTest.java]: done

**Comment by tonyshenkk** [comment]
done @normanmaurer

**Comment by Scottmitch** [comment]
@tonyshenkk - can you please sign Netty's [ICLA](http://netty.io/s/icla)?

**Comment by tonyshenkk** [comment]
@Scottmitch signed

**Comment by Scottmitch** [comment]
4.1 (c8a23028a)

**Comment by Scottmitch** [comment]
thanks @tonyshenkk !

**Comment by tonyshenkk** [comment]
@Scottmitch  It's my pleasure.

**Comment by normanmaurer** [comment]
@tonyshenkk @Scottmitch do we also need to backport to 4.0 tho ?

**Comment by tonyshenkk** [comment]
@normanmaurer yes, there is the same bug in 4.0.49.Final release version.

**Comment by normanmaurer** [comment]
@tonyshenkk ok will take care.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (4764d12)
